lib/client.js

```diff
@@ -319,9 +606,205 @@ MatrixClient.prototype.sendEvent = function(roomId, eventType, content, txnId,
+                    sessionIds.push(sessionId);
+                }
+            }
+            // Use the session with the lowest ID.
+            sessionIds.sort();
+            if (sessionIds.length === 0) {
+                // If we don't have a session for a device then
+                // we can't encrypt a message for it.
+                continue;
+            }
+            sessionId = sessionIds[0];
+            var session = new Olm.Session();
+            try {
+                session.unpickle(client.accountKey, sessions[sessionId]);
+                ciphertext[deviceKey] = session.encrypt(payloadString);
+                var pickled = session.pickle(client.accountKey);
+                client.sessionStore.storeEndToEndSession(
+                    deviceKey, sessionId, pickled
+                );
+            } finally {
+                session.free();
+            }
+        }
+        var encryptedContent = {
+            algorithm: e2eRoomInfo.algorithm,
+            sender_key: client.deviceCurve25519Key,
+            ciphertext: ciphertext
+        };
+        return encryptedContent;
+    } else {
+        throw new Error("Unknown end-to-end algorithm");
+    }
+}
+
+function _decryptMessage(client, event) {
+    if (client.sessionStore === null) {
+        // End to end encryption isn't enabled if we don't have a session
+        // store.
+        return _badEncryptedMessage(event, "Encryption not enabled");
+    }
+
+    var content = event.getContent();
+    if (content.algorithm === OLM_ALGORITHM) {
+        var deviceKey = content.sender_key;
+        var ciphertext = content.ciphertext;
+
+        if (!ciphertext) {
+            return _badEncryptedMessage(event, "Missing ciphertext");
+        }
+        if (!(client.deviceCurve25519Key in content.ciphertext)) {
+            return _badEncryptedMessage(event, "Not included in recipients");
+        }
+        var message = content.ciphertext[client.deviceCurve25519Key];
+        var sessions = client.sessionStore.getEndToEndSessions(deviceKey);
+        var payloadString = null;
+        var foundSession = false;
+        var session;
+        for (var sessionId in sessions) {
+            if (sessions.hasOwnProperty(sessionId)) {
+                session = new Olm.Session();
+                try {
+                    session.unpickle(client.accountKey, sessions[sessionId]);
+                    if (message.type === 0 && session.matches_inbound(message.body)) {
+                        foundSession = true;
+                    }
+                    payloadString = session.decrypt(message.type, message.body);
+                    var pickled = session.pickle(client.accountKey);
+                    client.sessionStore.storeEndToEndSession(
+                        deviceKey, sessionId, pickled
+                    );
+                } catch (e) {
+                    // Failed to decrypt with an existing session.
+                    console.log(
```

Comment:
I would guess this should probably rethrow or possibly emit a 'bad encrypted message' event rather than logging a console message and just not doing anything.